crates/simulations/src/cozy/agents/supplier.rs

```diff
@@ -0,0 +1,178 @@
+    risk_model: SetRiskModel,
+    risk_aversion: f64,
+    total_wealth: U256,
+    reaction_time: ReactionTime,
+}
+
+impl SupplierPreferences {
+    pub fn new(
+        risk_model: SetRiskModel,
+        risk_aversion: f64,
+        total_wealth: U256,
+        reaction_time: ReactionTime,
+    ) -> Self {
+        Self {
+            risk_model,
+            risk_aversion,
+            total_wealth,
+            reaction_time,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Supplier {
+    _name: Cow<'static, str>,
+    address: Address,
+    protocol: Arc<ProtocolContracts>,
+    set: Arc<SetContracts>,
+    preferences: SupplierPreferences,
+    rng: StdRng,
+}
+
+impl Supplier {
+    pub fn new(
+        _name: Cow<'static, str>,
+        address: Address,
+        protocol: Arc<ProtocolContracts>,
+        set: Arc<SetContracts>,
+        preferences: SupplierPreferences,
+        rng: StdRng,
+    ) -> Self {
+        Self {
+            _name,
+            address,
+            protocol,
+            set,
+            preferences,
+            rng,
+        }
+    }
+}
+
+impl Agent<CozyUpdate, CozyWorld> for Supplier {
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    fn activation_step(&mut self, state: &mut State<CozyUpdate, CozyWorld>) {
+        let router_approve_tx = self
+            .set
+            .base_token
+            .approve(self.protocol.cozy_router.address(), U256::MAX);
+        let _ = state.execute_evm_tx_and_decode(self.address, router_approve_tx);
+    }
+
+    fn step(
+        &mut self,
+        state: &State<CozyUpdate, CozyWorld>,
+        channel: &AgentChannelSender<CozyUpdate>,
+    ) {
+        if !self
+            .preferences
+            .reaction_time
+            .time_to_react(state.timestamp(), &mut self.rng)
+        {
+            return;
+        }
+
+        // Get current balance.
+        let balance = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.base_token.balance_of(self.address.into()),
+            )
+            .expect("Error getting balance.");
+        let current_set_shares = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.set.balance_of_matured(self.address.into()),
+            )
+            .expect("Error getting shares.");
+        let current_value = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.set.convert_to_assets(current_set_shares),
+            )
+            .expect("Error converting shares to assets.");
+        self.preferences.total_wealth = balance + current_value;
+
+        // Compute optimal allocation.
+        let optimal_allocation = self.compute_optimal_allocation(state);
+        let optimal_allocation_value =
+            f64_to_u256(optimal_allocation) * self.preferences.total_wealth;
+
+        // Supply or withdraw to target optimal allocation.
```

Comment:
do we consider gas fees at all? what is the gas fee set to anyway

Reply:
No considerations of gas fees yet, although would not be hard to add down the line.

Currently, sim assumes [gas fees are zero](https://github.com/Cozy-Finance/agent-buttercup/pull/26/files#diff-5f58d20cc74c20ee7605c6f90548f211ca6c5f1993da6741b09c665b4878671eR10).